ktor-network/jvm/src/io/ktor/network/sockets/DatagramSendChannel.kt

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.sockets
+
+import io.ktor.network.selector.*
+import io.ktor.network.util.*
+import io.ktor.utils.io.core.*
+import io.ktor.utils.io.pool.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.selects.*
+import kotlinx.coroutines.sync.*
+import java.net.*
+import java.nio.*
+import java.nio.channels.*
+
+internal class DatagramSendChannel(
+    val channel: DatagramChannel,
+    val socket: DatagramSocketImpl
+) : SendChannel<Datagram> {
+    @ExperimentalCoroutinesApi
+    override val isClosedForSend: Boolean
+        get() = socket.isClosed
+
+    @ExperimentalCoroutinesApi
+    override val isFull: Boolean
+        get() = if (isClosedForSend) false else lock.isLocked
+
+    private val lock = Mutex()
+
+    override fun close(cause: Throwable?): Boolean {
+        if (socket.isClosed) {
+            return false
+        }
+
+        socket.close()
+        return true
+    }
+
+
+    override fun offer(element: Datagram): Boolean {
+        if (!lock.tryLock()) return false
+
+        try {
+            DefaultDatagramByteBufferPool.useInstance { buffer ->
+                element.writeMessageTo(buffer)
+                return channel.send(buffer, element.address) != 0
+            }
+        } finally {
+            lock.unlock()
+        }
+    }
+
+    override suspend fun send(element: Datagram) {
+        lock.withLock {
+            DefaultDatagramByteBufferPool.useInstance { buffer ->
+                element.writeMessageTo(buffer)
+
+                val rc = channel.send(buffer, element.address)
+                if (rc != 0) {
+                    socket.interestOp(SelectInterest.WRITE, false)
+                    return
+                }
+
+                sendSuspend(buffer, element.address)
+            }
+
+        }
+    }
+
+    private suspend fun sendSuspend(buffer: ByteBuffer, address: SocketAddress) {
+        while (true) {
+            socket.interestOp(SelectInterest.WRITE, true)
+            socket.selector.select(socket, SelectInterest.WRITE)
+
+            if (channel.send(buffer, address) != 0) {
+                socket.interestOp(SelectInterest.WRITE, false)
+                break
+            }
+        }
+    }
+
+    override val onSend: SelectClause2<Datagram, SendChannel<Datagram>>
+        get() = TODO("[DatagramSendChannel] doesn't support [onSend] select clause")
+
+    @ExperimentalCoroutinesApi
+    override fun invokeOnClose(handler: (cause: Throwable?) -> Unit) {
+        TODO("[DatagramSendChannel] doesn't support [invokeOnClose] operation.")
```

Comment:
Unlike onSend, invokeOnClose seems to be useful. Some kxcoroutine parts depend on it. Can't we support it? Or, at least, a single handler registration?

Reply:
Thanks for the catch! Implemented

Reply:
Looks great so far, here is what I would change: https://github.com/Thomas-Vos/ktor/commit/184a61848d0244cb6773685c93e3e05133c3920f

Reply:
Hey @Thomas-Vos, thanks for the suggestion. The tests are great. One minor thing I found: there is a race with `cause` value in the suggestion(it's possible to find `null` cause(in `invokeOnClose`) before the cause was actually set(in `close`). I will take your suggestions with minor fixes. Thanks again.